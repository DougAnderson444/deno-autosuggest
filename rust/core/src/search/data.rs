use std::borrow::Cow;
use std::default::Default;
use super::Scores;
use crate::lexis::{Text, tokenize_record};


#[derive(Clone, Debug)]
pub struct Record {
    id: usize,
    text: Text<Vec<char>>,
}


impl Record {
    pub fn new<T: IntoIterator<Item=char>>(id: usize, source: T) -> Record {
        let source: Vec<char> = source.into_iter().collect();
        Record {
            id,
            text: tokenize_record(&source).to_owned(),
        }
    }

    pub fn to_hit<'a>(&'a self) -> Hit<Cow<'a, [char]>> {
        Hit::new(self.id, self.text.to_cow())
    }
}


#[derive(Debug, Clone)]
pub struct Hit<T: AsRef<[char]>> {
    pub id:      usize,
    pub text:    Text<T>,
    pub scores:  Scores,
}


impl<T: AsRef<[char]>> Hit<T> {
    pub fn new(id: usize, text: Text<T>) -> Hit<T> {
        Hit { id, text, scores: Default::default() }
    }
}


#[derive(Debug, Clone)]
pub struct SearchResult {
    pub id:          usize,
    pub highlighted: Vec<char>,
}
