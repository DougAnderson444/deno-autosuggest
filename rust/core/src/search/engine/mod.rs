mod score;

use std::cmp::Ordering;
use crate::lexis::Text;
use super::{Record, Hit};
use score::score;
pub use score::Scores;


pub struct Engine<'a> {
    query: Text<&'a [char]>,
    hits:  Vec<Hit<'a>>,
    limit: usize,
}


impl<'a> Engine<'a> {
    pub fn new(query: Text<&'a [char]>, limit: usize) -> Self {
        Self {
            query,
            hits: Vec::with_capacity(limit * 2),
            limit,
        }
    }

    pub fn push(&mut self, record: &'a Record) {
        let mut hit = Hit::from_record(record);
        score(&self.query, &mut hit);
        if self.matches(&hit) {
            self.hits.push(hit);
        }
        if self.hits.len() >= self.limit * 2 {
            self.sort_and_truncate();
        }
    }

    pub fn push_many<I: IntoIterator<Item=&'a Record>>(&mut self, records: I) {
        for record in records {
            self.push(record);
        }
    }

    pub fn matches(&self, hit: &Hit<'a>) -> bool {
        let matches = &hit.scores.matches;
        if self.query.is_empty() { return true; }
        if matches.len() == 0 { return false; }
        if matches.len() == 1 && self.query.words.len() > 1 {
            let word_match = &matches[0];
            let unfinished = !word_match.fin;
            let first_half = (word_match.query.len * 2) < word_match.record.len;
            if unfinished && first_half { return false; }
        }
        true
    }

    pub fn sort_and_truncate(&mut self) {
        self.hits.sort_by(|hit1, hit2| {
            hit1.scores.linear.iter()
                .zip(hit2.scores.linear.iter())
                .map(|(s1, s2)| s2.cmp(s1))
                .find(|&ord| ord != Ordering::Equal)
                .unwrap_or(Ordering::Equal)
        });
        self.hits.truncate(self.limit);
    }

    pub fn hits<'b>(&'b self) -> &'b [Hit<'a>] {
        &self.hits
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::tokenize_query;
    use super::{Record, Engine};

    fn check(name: &str, queries: &[&str]) {
        let records = [
            Record::new(10, "brown plush bear"),
            Record::new(20, "metal detector"),
            Record::new(30, "yellow metal mailbox"),
        ];
        for (i, query) in queries.iter().enumerate() {
            let query = tokenize_query(query);
            let mut finder = Engine::new(query.to_ref(), 10);
            finder.push_many(&records);
            finder.sort_and_truncate();
            assert_debug_snapshot!(format!("{}-{}", name, i), finder.hits);
        }
    }

    #[test]
    fn search_empty() {
        check("empty", &[""]);
    }

    #[test]
    fn search_equal() {
        check("equal", &["yelow metall maiblox"]);
    }

    #[test]
    fn search_partial() {
        check("partial", &[
            "brown plush bear",
            "metal detector",
            "yellow metal mailbox",
        ]);
    }

    #[test]
    fn search_intersection() {
        check("intersection", &[
            "red wooden mailbox",
            "red wooden mail",
        ]);
    }

    #[test]
    fn search_min_match() {
        check("min_match", &[
            "wooden mai",
            "wooden mail",
        ]);
    }

    #[test]
    fn search_transpositions() {
        check("transpositions", &[
            "metal mailbox",
            "mailbox metal",
        ]);
    }
}
