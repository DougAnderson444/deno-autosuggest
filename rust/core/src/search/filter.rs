use crate::tokenization::TextRef;
use crate::search::Hit;

pub fn hit_matches(query: &TextRef, hit: &Hit) -> bool {
    if query.is_empty() { return true; }
    if hit.rmatches.len() == 0 { return false; }
    if hit.rmatches.len() == 1 && hit.qmatches.len() == 1 && query.words.len() > 1 {
        let rmatch     = &hit.rmatches[0];
        let qmatch     = &hit.qmatches[0];
        let first_half = (qmatch.word_len() * 2) < rmatch.word_len();
        if !rmatch.fin && first_half { return false; }
    }
    true
}
