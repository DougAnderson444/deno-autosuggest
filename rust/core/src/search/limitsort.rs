use std::cmp::Ordering;
use crate::search::Hit;


pub struct LimitSort<'a, Src: Iterator<Item=Hit<'a>>> {
    source: Src,
    buffer: Vec<Hit<'a>>,
    limit:  usize,
    done:   bool,
}


impl<'a, Src: Iterator<Item=Hit<'a>>> LimitSort<'a, Src> {
    pub fn new(source: Src, limit: usize) -> Self {
        Self {
            source,
            buffer: Vec::with_capacity(limit * 2),
            limit,
            done: false,
        }
    }

    fn sort_and_truncate(&mut self) {
        self.buffer.sort_by(|hit1, hit2| {
            hit1.scores.linear.iter()
                .zip(hit2.scores.linear.iter())
                .map(|(s1, s2)| s2.cmp(s1))
                .find(|&ord| ord != Ordering::Equal)
                .unwrap_or(Ordering::Equal)
        });
        self.buffer.truncate(self.limit);
    }
}


impl<'a, Src: Iterator<Item=Hit<'a>>> Iterator for LimitSort<'a, Src> {
    type Item = Hit<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if !self.done {
            while let Some(hit) = self.source.next() {
                self.buffer.push(hit);
                if self.buffer.len() >= self.limit * 2 {
                    self.sort_and_truncate();
                }
            }
            self.sort_and_truncate();
            self.buffer.reverse();
            self.done = true;
        }

        self.buffer.pop()
    }
}
