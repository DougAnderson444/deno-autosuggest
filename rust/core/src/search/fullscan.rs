use crate::store::Record;
use crate::search::Hit;


pub struct FullScan<'a, Src: Iterator<Item=&'a Record>> {
    source: Src,
}


impl<'a, Src: Iterator<Item=&'a Record>> FullScan<'a, Src> {
    pub fn new(source: Src) -> Self {
        Self { source }
    }
}


impl<'a, Src: Iterator<Item=&'a Record>> Iterator for FullScan<'a, Src> {
    type Item = Hit<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.source
            .next()
            .map(Hit::from_record)
    }
}
