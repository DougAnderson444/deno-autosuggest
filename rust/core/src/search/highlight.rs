use crate::tokenization::Text;
use crate::matching::WordMatch;
use crate::search::Hit;


pub fn highlight(hit: &Hit, dividers: (&[char], &[char])) -> String {
    let (div_left, div_right) = dividers;
    let Hit {
        title: Text { words, source, .. },
        matches,
        ..
    } = hit;

    let mut highlighted = {
        let chars_src = source.len();
        let chars_hl  = (div_left.len() + div_right.len() + 1) * words.len();
        String::with_capacity((chars_src + chars_hl) * 4)
    };

    let mut offset = 0;
    for (ix, word) in words.iter().enumerate() {
        match matches.iter().find(|m| m.record.ix == ix) {
            Some(WordMatch { record: m, .. }) => {
                let match_start = word.place.0 + m.slice.0;
                let match_end   = word.place.0 + m.slice.1;
                highlighted.extend(&source[offset .. match_start]);
                highlighted.extend(div_left);
                highlighted.extend(&source[match_start .. match_end]);
                highlighted.extend(div_right);
                highlighted.extend(&source[match_end .. word.place.1]);
            },
            None => {
                highlighted.extend(&source[offset .. word.place.1]);
            },
        }
        offset = word.place.1;
    }
    highlighted.extend(&source[offset .. ]);
    highlighted.retain(|ch| ch != '\0');

    highlighted
}


#[cfg(test)]
mod tests {
    use crate::matching::{WordMatch, MatchSide};
    use crate::store::Record;
    use crate::search::Hit;
    use crate::lang::{lang_german, lang_portuguese};
    use super::highlight;

    const L: &[char] = &['['];
    const R: &[char] = &[']'];

    fn mock_match(ix: usize, size: usize) -> WordMatch {
        WordMatch {
            query:  MatchSide { ix:  0, len: size, slice: (0, size), function: false },
            record: MatchSide { ix: ix, len: size, slice: (0, size), function: false },
            typos:  0,
            fin:    false,
        }
    }

    #[test]
    fn highlight_basic() {
        let record = Record::new(10, "metal detector", 0, &None);

        let mut hit = Hit::from_record(&record);
        hit.matches.push(mock_match(1, 6));

        let expected = "metal [detect]or";
        let received = highlight(&hit, (L, R));

        assert_eq!(&received, expected);
    }

    #[test]
    fn highlight_stripped() {
        let record = Record::new(10, "'metal' mailbox!", 0, &None);

        let mut hit = Hit::from_record(&record);
        hit.matches.push(mock_match(0, 5));

        let expected = "'[metal]' mailbox!";
        let received = highlight(&hit, (L, R));

        assert_eq!(&received, expected);
    }

    #[test]
    fn highlight_multichar_dividers() {
        let record = Record::new(10, "metal detector", 0, &None);

        let mut hit = Hit::from_record(&record);
        hit.matches.push(mock_match(1, 6));

        let l: &[char] = &['{', '{'];
        let r: &[char] = &['}', '}'];

        let expected = "metal {{detect}}or";
        let received = highlight(&hit, (l, r));

        assert_eq!(&received, expected);
    }

    #[test]
    fn highlight_utf_padded() {
        let record = Record::new(10, "Passstraße", 0, &Some(lang_german()));

        let mut hit = Hit::from_record(&record);
        hit.matches.push(mock_match(0, 9));

        let expected = "[Passstraß]e";
        let received = highlight(&hit, (L, R));

        assert_eq!(&received, expected);
    }

    #[test]
    fn highlight_utf_nfd() {
        let record = Record::new(10, "Passstraße", 0, &Some(lang_portuguese()));

        let mut hit = Hit::from_record(&record);
        hit.matches.push(mock_match(0, 9));

        let expected = "[Passstraß]e";
        let received = highlight(&hit, (L, R));

        assert_eq!(&received, expected);
    }
}
