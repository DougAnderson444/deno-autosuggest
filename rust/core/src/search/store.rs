use crate::lexis::tokenize_query;
use super::data::{Record, SearchResult};
use super::engine::Engine;
use super::highlight::highlight;


static DEFAULT_LIMIT: usize = 10;


pub struct Store {
    records:    Vec<Record>,
    separators: (Vec<char>, Vec<char>),
    limit:      usize,
    results:    Vec<SearchResult>,
}


impl Store {
    pub fn new() -> Self {
        Self {
            records:    Vec::new(),
            separators: (vec!['['], vec![']']),
            limit:      DEFAULT_LIMIT,
            results:    Vec::with_capacity(DEFAULT_LIMIT),
        }
    }

    pub fn add(&mut self, record: Record) {
        self.records.push(record);
    }

    pub fn highlight_using(&mut self, separators: (&str, &str)) {
        let left:  Vec<char> = separators.0.chars().collect();
        let right: Vec<char> = separators.1.chars().collect();
        self.separators = (left, right);
    }

    pub fn results<'a>(&'a self) -> &'a [SearchResult] {
        &self.results
    }

    pub fn search(&mut self, query: &str) {
        let query: Vec<char> = query.chars().collect();
        let query = tokenize_query(&query);

        let mut engine = Engine::new(&query, self.limit);
        engine.push_many(&self.records);
        engine.sort_and_truncate();

        let separators = (&self.separators.0[..], &self.separators.1[..]);
        let results = &mut self.results;
        results.clear();
        results.extend(
            engine.hits()
                .iter()
                .map(|hit| {
                    SearchResult {
                        id: hit.id,
                        highlighted: highlight(hit, separators)
                    }
                })
        );
    }
}
