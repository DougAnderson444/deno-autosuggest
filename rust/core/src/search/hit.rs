use std::default::Default;
use crate::tokenization::TextRef;
use crate::matching::WordMatch;
use crate::store::Record;
use super::score::Scores;


#[derive(Debug)]
pub struct Hit<'a> {
    pub id:      usize,
    pub title:   TextRef<'a>,
    pub rating:  usize,
    pub rmatches: Vec<WordMatch>,
    pub qmatches: Vec<WordMatch>,
    pub scores:  Scores,
}


impl<'a> Hit<'a> {
    pub fn from_record(record: &'a Record) -> Hit<'a> {
        Hit {
            id:       record.id,
            title:    record.title.to_ref(),
            rating:   record.rating,
            scores:   Default::default(),
            rmatches: Vec::new(),
            qmatches: Vec::new(),
        }
    }
}
