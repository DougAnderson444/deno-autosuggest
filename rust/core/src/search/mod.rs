mod score;
mod filter;
mod sort;
mod highlight;

use std::default::Default;
use crate::utils::LimitSortIterator;
use crate::tokenization::Text;
use crate::matching::WordMatch;
use crate::store::{Store, Record};


#[derive(Debug)]
pub struct Hit<'a> {
    pub id:      usize,
    pub title:   Text<&'a [char]>,
    pub rating:  usize,
    pub matches: Vec<WordMatch>,
    pub scores:  Scores,
}


impl<'a> Hit<'a> {
    pub fn from_record(record: &'a Record) -> Hit<'a> {
        Hit {
            id:      record.id,
            title:   record.title.to_ref(),
            rating:  record.rating,
            scores:  Default::default(),
            matches: Vec::new(),
        }
    }
}


const SCORES_SIZE: usize = 9;


pub enum ScoreType {
    SameWords   = 0,
    SamePrimary = 1,
    Typos       = 2,
    Trans       = 3,
    Fin         = 4,
    Offset      = 5,
    Rating      = 6,
    WordLen     = 7,
    CharLen     = 8,
}


#[derive(Debug, Clone)]
pub struct Scores([isize; SCORES_SIZE]);


impl Scores {
    pub fn iter(&self) -> impl Iterator<Item=&isize> {
        self.0.iter()
    }
}


impl std::ops::Index<ScoreType> for Scores {
    type Output = isize;

    fn index(&self, score: ScoreType) -> &Self::Output {
        &self.0[score as usize]
    }
}


impl std::ops::IndexMut<ScoreType> for Scores {
    fn index_mut(&mut self, score: ScoreType) -> &mut Self::Output {
        &mut self.0[score as usize]
    }
}


impl Default for Scores {
    fn default() -> Scores {
        Scores([0; SCORES_SIZE])
    }
}


#[derive(Debug)]
pub struct SearchResult {
    pub id:    usize,
    pub title: String,
}


pub fn search<'a>(
    store:   &'a Store,
    query:   &'a Text<&'a [char]>,
) -> std::iter::Map<impl Iterator<Item=Hit<'a>>, impl (FnMut(Hit<'a>) -> SearchResult)> {
    let dividers = store.dividers();
    store.records.iter()
        .map(|record| {
            Hit::from_record(record)
        })
        .map(move |mut hit| {
            score::score(query, &mut hit);
            hit
        })
        .filter(move |hit| {
            filter::hit_matches(query, hit)
        })
        .limit_sort(store.limit, sort::compare_hits)
        .map(move |hit| {
            SearchResult {
                id:    hit.id,
                title: highlight::highlight(&hit, dividers),
            }
        })
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::tokenize_query;
    use crate::lang::{Lang, lang_english};
    use crate::store::{Store, Record};
    use super::search;

    fn check(name: &str, lang: Option<Lang>, queries: &[&str]) {
        let mut store = Store::new();
        store.lang = lang;
        store.add(Record::new(10, "brown plush bear",     10, &store.lang));
        store.add(Record::new(20, "the metal detector",   20, &store.lang));
        store.add(Record::new(30, "yellow metal mailbox", 30, &store.lang));
        store.add(Record::new(40, "thesaurus",            40, &store.lang));
        store.add(Record::new(50, "wi-fi router",         50, &store.lang));

        for (i, query) in queries.iter().enumerate() {
            let query   = tokenize_query(query, &store.lang);
            let query   = query.to_ref();
            let results = search(&store, &query).collect::<Vec<_>>();
            assert_debug_snapshot!(format!("{}-{}", name, i), results);
        }
    }

    #[test]
    fn search_empty() {
        check("empty", None, &[""]);
    }

    #[test]
    fn search_equal() {
        check("equal", None, &["yelow metall maiblox"]);
    }

    #[test]
    fn search_partial() {
        check("partial", None, &[
            "brown plush bear",
            "metal detector",
            "yellow metal mailbox",
        ]);
    }

    #[test]
    fn search_intersection() {
        check("intersection", None, &[
            "red wooden mailbox",
            "red wooden mail",
        ]);
    }

    #[test]
    fn search_min_match() {
        check("min_match", None, &[
            "wooden mai",
            "wooden mail",
        ]);
    }

    #[test]
    fn search_transpositions() {
        check("transpositions", None, &[
            "metal mailbox",
            "mailbox metal",
        ]);
    }


    #[test]
    fn search_stemming() {
        let mut store = Store::new();
        store.lang = Some(lang_english());
        store.add(Record::new(30, "universe", 30, &store.lang));

        let query1   = tokenize_query("university", &None);
        let query2   = tokenize_query("university", &store.lang);
        let query1   = query1.to_ref();
        let query2   = query2.to_ref();
        let results1 = search(&store, &query1).collect::<Vec<_>>();
        let results2 = search(&store, &query2).collect::<Vec<_>>();

        assert_debug_snapshot!(results1);
        assert_debug_snapshot!(results2);
    }

    #[test]
    fn search_particles() {
        check("particles_nolang", None, &[
            "the",
        ]);

        check("particles", Some(lang_english()), &[
            "the",
        ]);
    }

    #[test]
    fn search_joined() {
        check("joined_query", None, &[
            "wifi",
        ]);

        check("joined_record", None, &[
            "the saurus",
        ]);
    }
}
