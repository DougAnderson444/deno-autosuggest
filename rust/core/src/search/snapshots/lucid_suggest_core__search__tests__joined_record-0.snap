---
source: src/search/mod.rs
expression: results
---
[
    SearchResult {
        id: 40,
        title: "[thesaurus]",
    },
    SearchResult {
        id: 20,
        title: "[the] metal detector",
    },
]
